internal/pkg/stream/stream.go

```diff
@@ -48,34 +47,34 @@ func Stream(ctx context.Context, streamer Streamer) error {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
-		case <-streamer.Done():
-			// No more events to Fetch, flush and exit successfully.
-			streamer.Notify()
-			return nil
 		case <-time.After(fetchDelay):
-			next, err = streamer.Fetch()
+			next, done, err = streamer.Fetch()
 			if err != nil {
 				return err
 			}
 			streamer.Notify()
+			if done {
+				return nil
+			}
 		}
 	}
 }
 
 // nextFetchDate returns a time to wait using random jitter and exponential backoff.
 func nextFetchDate(clock clock, rand func(int) int, retries int) time.Time {
-	// waitMs := rand.Intn( 							// Get a random integer between 0 and ...
+	// waitMs := rand.Intn( 							// Get a random integer between streamerMinFetchIntervalDurationMs and ...
 	// 	min( 											// the minimum of ...
 	// 		streamerMaxFetchIntervalDuration,           // the max fetch interval and ...
 	// 		streamerFetchIntervalDuration*(1<<retries), // d*2^r, where r=retries and d= the normal
-	// 	),
-	// )
-	waitMs := rand(
-		min(
-			streamerMaxFetchIntervalDurationMs,
-			streamerFetchIntervalDurationMs*(1<<retries),
-		),
-	)
+	// 	)-streamerMinFetchIntervalDurationMs
+	// ) + streamerMinFetchIntervalDurationMs
+	waitMs :=
+		rand(
+			min(
+				streamerMaxFetchIntervalDurationMs,
+				streamerFetchIntervalDurationMs*(1<<retries),
+			)-streamerMinFetchIntervalDurationMs,
```

Comment:
if min generated a number less than `streamerMinFetchIntervalDurationMs`, `rand` would get passed a negative number and panic - unless i'm reading that wrong?

Reply:
min can't generate a number less than `streamerMinFetchIntervalDurationMs`. so it will be

`[0, min(streamerMaxFetchIntervalDurationMs, streamerFetchIntervalDurationMs*(1<<retries))-streamerMinFetchIntervalDurationMs)` for the `rand` and then
`[streamerMinFetchIntervalDurationMs, min(streamerMaxFetchIntervalDurationMs, streamerFetchIntervalDurationMs*(1<<retries)))` after the addition.

Reply:
ahhh yeah i gotcha, i was thinking `rand` logic on the inside instead of min.👍